EventFahrplan/EventFahrplan
# Support scheduling exact alarms for sessions.

# Description
+ Add `android.permission.SCHEDULE_EXACT_ALARM`.
+ Extract common code into `SessionAlarmViewModelDelegate`.
+ Revise alarm related names, hence harmonize `SessionDetailsViewModel` & `FahrplanViewModel`.
+ This change does not affect alarms scheduled for loading schedule updates.

# Test scenarios
+ Add alarm via:
  + Main schedule grid screen: long press a session
  + Session details screen: toolbar item
+ Fresh install.
+ Disabling notifications.
+ Revoking schedule exact alarm permission.
+ Check correct error messages.
+ [Doze mode](https://developer.android.com/training/monitoring-device-state/doze-standby#testing_doze), successfully tested with Android 14 device
+ [App Standby mode](https://developer.android.com/training/monitoring-device-state/doze-standby#testing_your_app_with_app_standby), successfully tested with Android 14 device

# Successfully tested on
with `ccc37c3` flavor, `debug` build
- :heavy_check_mark: Pixel 6 emulator, Android 10 (API 29)
- :heavy_check_mark: Pixel 3a device, Android 12 (API 31)
- :heavy_check_mark: Pixel 4a emulator, Android 12L (API 32)
- :heavy_check_mark: Pixel 7 emulator, Android 13 (API 33)
- :heavy_check_mark: Pixel 6 device, Android 14 (API 34)

# Screenshots (Android 14)
![Notification permission dialog](https://github.com/EventFahrplan/EventFahrplan/assets/144518/3123dc4e-e5cd-43e7-ad16-2dd8b5046bf5) ![Allow setting alarms and reminders screen](https://github.com/EventFahrplan/EventFahrplan/assets/144518/959b13e8-fe5b-4360-90a4-2231f7ea7bda) ![setting alarms and reminders disabled error message](https://github.com/EventFahrplan/EventFahrplan/assets/144518/dd3ff388-14e6-4ce6-8913-5888e30f4fbe)


# Docs
+ https://developer.android.com/about/versions/12/behavior-changes-12#exact-alarm-permission
+ https://developer.android.com/about/versions/13/features#use-exact-alarm-permission
+ https://developer.android.com/about/versions/14/changes/schedule-exact-alarms
+ https://developer.android.com/reference/android/Manifest.permission#SCHEDULE_EXACT_ALARM
+ https://developer.android.com/reference/androidx/core/app/AlarmManagerCompat#setExactAndAllowWhileIdle(android.app.AlarmManager,int,long,android.app.PendingIntent)
+ https://developer.android.com/reference/android/app/AlarmManager#canScheduleExactAlarms()
+ https://developer.android.com/training/monitoring-device-state/doze-standby#testing_doze_and_app_standby

---

Resolves #612
